server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java

```diff
@@ -99,15 +99,23 @@ public DiskThresholdMonitor(Settings settings, Supplier<ClusterState> clusterSta
         this.client = client;
     }
 
+    private boolean checkInProgress(final boolean expectedValue, final boolean newValue) {
+        if (diskThresholdSettings.allowForcedUpdate()) {
+            assert checkInProgress.get() == false;
+            return true;
+        }
+        return checkInProgress.compareAndSet(expectedValue, newValue);
+    }
+
     private void checkFinished() {
-        final boolean checkFinished = checkInProgress.compareAndSet(true, false);
+        final boolean checkFinished = checkInProgress(true, false);
         assert checkFinished;
         logger.trace("checkFinished");
     }
 
     public void onNewInfo(ClusterInfo info) {
 
-        if (checkInProgress.compareAndSet(false, true) == false) {
+        if (checkInProgress(false, true) == false) {
```

Comment:
Aren't we potentially introducing all kinds of races here when we run this concurrently and update a bunch of maps in parallel? It seems that the logic in this method doesn't really expect this to happen and it could actually have functional impact with all the various checks we do on those maps in the computations below?

Reply:
Yes, but that's the only "simple" way I found to ensure that DiskThresholdMonitor effectively runs the reroute after the disk space has been updated... At this point I think we should just re execute refreshDiskUsage in the assertBusy() loop until we have what we want (as initially done in https://github.com/elastic/elasticsearch/pull/62358#discussion_r488598448).

Reply:
> At this point I think we should just re execute refreshDiskUsage in the assertBusy() loop until we have what we want (as initially done in #62358 (comment)).

Yea I tend to agree :( (begrudgingly :D ... maybe add a TODO to look into a cleaner way and do that?)

Reply:
I pushed 7cddec2a83bddac8bb12b26c0543398efc0a89e5